core/src/main/java/tech/tablesaw/io/xlsx/XlsxReader.java

```diff
@@ -0,0 +1,318 @@
+                    break;
+                }
+            }
+        }
+        return (col1 >= 0 && col2 >= col1 ? new int[] { col1, col2 } : null);
+    }
+
+    private InputStream getInputStream(final XlsxReadOptions options, final byte[] bytes) throws FileNotFoundException {
+        if (bytes != null) {
+            return new ByteArrayInputStream(bytes);
+        }
+        if (options.inputStream() != null) {
+            return options.inputStream();
+        }
+        return new FileInputStream(options.file());
+    }
+
+    private Table createTable(final Sheet sheet, final int[] tableArea, final XlsxReadOptions options) {
+        // assume header row if all cells are of type String
+        Row row = sheet.getRow(tableArea[0]);
+        final List<String> headerNames = new ArrayList<>();
+        for (final Cell cell : row) {
+            if (cell.getCellType() == CellType.STRING) {
+                headerNames.add(cell.getRichStringCellValue().getString());
+            } else {
+                break;
+            }
+        }
+        if (headerNames.size() == tableArea[3] - tableArea[2] + 1) {
+            tableArea[0]++;
+        } else {
+            headerNames.clear();
+            for (int col = tableArea[2]; col <= tableArea[3]; col++) {
+                headerNames.add("col" + col);
+            }
+        }
+        final Table table = Table.create(options.tableName());
+        final List<Column<?>> columns = new ArrayList<>();
+        while (columns.size() < headerNames.size()) {
+            columns.add(null);
+        }
+        for (int rowNum = tableArea[0]; rowNum <= tableArea[1]; rowNum++) {
+            row = sheet.getRow(rowNum);
+            for (int colNum = 0; colNum < headerNames.size(); colNum++) {
+                final Cell cell = row.getCell(colNum + tableArea[2], MissingCellPolicy.RETURN_BLANK_AS_NULL);
+                Column<?> column = columns.get(colNum);
+                if (cell != null) {
+                    if (column == null) {
+                        column = createColumn(headerNames.get(colNum), cell);
+                        columns.set(colNum, column);
+                        while (column.size() < rowNum - tableArea[0]) {
+                            column.appendMissing();
+                        }
+                    }
+                    final Column<?> altColumn = appendValue(column, cell);
+                    if (altColumn != null && altColumn != column) {
+                        column = altColumn;
+                        columns.set(colNum, column);
+                    }
+                }
+                if (column != null) {
+                    while (column.size() <= rowNum - tableArea[0]) {
+                        column.appendMissing();
+                    }
+                }
+            }
+        }
+        while (columns.remove(null))
+            ;
+        table.addColumns(columns.toArray(new Column<?>[columns.size()]));
```

Comment:
It might be nicer to add an `addColumns(Collection<Column> columns)` method rather than having to convert to an array here